apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/parser/GraphQLDocumentParser.kt

```diff
@@ -387,38 +387,55 @@ class GraphQLDocumentParser(val schema: Schema, private val packageNameProvider:
   }
 
   private fun GraphQLParser.InlineFragmentContext.parse(
-      parentSelectionSet: GraphQLParser.SelectionSetContext,
-      parentSchemaType: Schema.Type
+      parentSchemaType: Schema.Type,
+      parentFields: ParseResult<List<Field>>
+
   ): ParseResult<InlineFragment> {
     val typeCondition = typeCondition().typeName().NAME().text
     val schemaType = schema[typeCondition] ?: throw ParseException(
         message = "Unknown type`$typeCondition}`",
         token = typeCondition().typeName().start
     )
 
-    if (!parentSchemaType.isAssignableFrom(schema = schema, other = schemaType)) {
+    if (!parentSchemaType.isAssignableFrom(other = schemaType, schema = schema)) {
       throw ParseException(
-          message = "Fragment cannot be spread here as objects of type `${parentSchemaType.name}` can never be of type `$typeCondition`",
+          message = "Fragment cannot be spread here as result can never be of type `$typeCondition`",
           token = typeCondition().typeName().start
       )
     }
 
-    val possibleTypes = when (schemaType) {
-      is Schema.Type.Interface -> schemaType.possibleTypes?.map { it.rawType.name!! } ?: emptyList()
-      is Schema.Type.Union -> schemaType.possibleTypes?.map { it.rawType.name!! } ?: emptyList()
-      else -> listOf(typeCondition)
-    }.distinct()
+    val decoratedParentFields = parentFields.let { (parentFields, usedTypes) ->
+      // if inline fragment conditional type contains the same field as parent type
+      // carry over meta info such as: `description`, `isDeprecated`, `deprecationReason`
+      val decoratedFields = parentFields.map { parentField ->
+        when (schemaType) {
+          is Schema.Type.Interface -> schemaType.fields?.find { it.name == parentField.fieldName }
+          is Schema.Type.Object -> schemaType.fields?.find { it.name == parentField.fieldName }
+          is Schema.Type.Union -> schemaType.fields?.find { it.name == parentField.fieldName }
+          else -> null
+        }?.let { field ->
+          parentField.copy(
+              description = field.description ?: "",
```

Comment:
You might want to keep the original value if null. Like `description = field.description ?: description`

Right now, this always overrides it with empty String if field description is null